fix(core): Implement proper retry backoff logic with jitter.

ceache opened the pull request:
New retry logic takes a maximum percentage off the canonical backoff,
ensure gradual predictable retries timings while still having a
controlable amount of jitter (re-introducing the `max_jitter` parameter)
to avoids swarming client retries.

Fix regression introduced in 60366d2c7910fc833991fad8e04bbe33817c0544
where retry/backoff logic produced only whole second (integer) retry
delays.

This produced inadequate retries on first retry and would generally not
work on fast network where sub miliseconds retries are desired.
Additionally, with high `max_delay` setting, as the range was always
spanning from 0 until the last delay, it would also produce extremely
random results with short delays following longer ones which is contrary
to the expected backoff logic.

ceache commented:
The issue is very easy to demonstrate:

``` python
import kazoo.retry

def retry_me():
    raise  kazoo.retry.ForceRetryError()

l = []
r = kazoo.retry.KazooRetry(max_tries=100, sleep_func=l.append)
r(retry_me)
```
Then inspect the distribution of the "sleep times" in `l`.

With the default settings (`delay` starts at 100ms, `backoff` multiplier of 2, `max_delay` of 60 sec), the first 4 retries are 50% selections between 0 sec and 1 sec (which is both too fast and too slow).
As the retries accumulate, the range expands until spanning between 0 and 60, producing completely random results not in line with the intent of a "gradual backoff mechanism".

More pathological cases can be produced with different settings such as setting `max_delay` to 1 sec, a.k.a. the "fast network" case.

In the proposed code, I am trying to follow the spirit of the PR in 60366d2, avoiding the clustering effect, while resolving the issues described above.
The `max_jitter` value is "de-deprecated" to mean the maximum percentage of the sleep time that can be "jittered". For example, with the default of 0.8 (80%) `max_jitter`, the sleep time will be randomly picked between T\*0.6, and T\*1.4, i.e a 80% spread. Furthermore, to minimize the clustering effect, the `backoff` multiplier is applied to the actual slept time, increasing the "distance" between different clients.

ceache commented on kazoo/retry.py:
```diff
@@ -134,25 +130,24 @@ def __call__(self, func, *args, **kwargs):
-                self._cur_delay = min(self._cur_delay * self.backoff,
-                                      self.max_delay)
+                self._cur_delay = min(sleeptime * self.backoff,
+                                      self.max_delay)
```
fixed

ceache commented on kazoo/retry.py:
```diff
@@ -134,25 +130,25 @@ def __call__(self, func, *args, **kwargs):
-                self._cur_delay = min(self._cur_delay * self.backoff,
+                self._cur_delay = min(sleeptime * self.backoff,
                                       self.max_delay)
+
```
fixed

jeffwidman commented:
Thanks @ceache your explanation makes sense to me. 

@andreasgudmundsson can you review as well? I'd like to hear your input as well since you spent a bunch of time thinking about this problem in https://github.com/python-zk/kazoo/pull/521

StephenSorriaux commented:
Gentle ping @andreasgudmundsson. If you have some time available, we really would like your input on this, thank you.

andreasgudmundsson commented:
Hi, the proposed patch looks good and it does fix the regression mentioned. Now I don't work with large zk installation anymore so I can't test this in a real-world setting and I don't have time to create a simulation. But here are my 2 cents(napkin calculations).

With a back-off factor of 2x and max-jitter of 100% we'll now see backoff intervals like

[1,3] ==  2 * [0.5, 1.5]
[2,6] == 4 * [0.5, 1.5]
[4,12] == 8 * [0.5, 1.5]

Uniform distrubution with mean $backoff and length $backoff. The n-th wave of reconnects will have to sleep  for n/2 seconds. Backoff intervals like [0, n**2] are not possible. Both types of intervals are achievable if jitter is calculated as 
  jitter := randint(100 - max_jitter, 100 + max_jitter),
instead of 
  jitter := randint(100 - max_jitter//2, 100 + max_jitter//2),

jeffwidman commented:
@ceache I just realized this was never merged... it'd be nice to land this. 

What do you think about @andreasgudmundsson proposed improvement?
>  Both types of intervals are achievable if jitter is calculated as
`jitter := randint(100 - max_jitter, 100 + max_jitter)`
instead of
`jitter := randint(100 - max_jitter//2, 100 + max_jitter//2)`

ceache commented:
Done,

Updated to not half range, i.e. `random.uniform(1.0-max_jitter, 1.0+max_jitter)`
Removed back and forth useless conversion to integer.

Cheers,

jeffwidman commented:
Looks like you also need to merge `master`?

ceache commented:
@jeffwidman rebased on latest master
